### PR TITLE
feat: admin dashboard implementation

### PR DESCRIPTION
**Pull request by purplena**
**Status for 01/07/2024**

1. Login/logout to/from admin section with validations
2. Widdlewear implementation to protect the routes (questions in the code)
3. Database connection works properly. Took me a lot of time to figure out the problems with mysql connection
4. Fixed problems with contact form. Due to Laravel update to a new version, needed to add `->to(config('admin.email'))` in the ContactMail.php 
5. Replaced env() method with config('admin.email') and created admin.php in config section

**Status for 02/07/2024**
1. Installed Laravel Pint. Added additional rules of formatting by creating `pint.json` file. The rules were taken from PHP-CS-Fixer  and include: `class_attributes_separation` (to leave one space between methods and attributes), `no_extra_blank_lines` (configured to leave no blanc space before and after curly braces in class methods), `no_empty_statement` (to remove extra semicolons). Updated `settings.json` file in VSCode to enable formatting on save. 
2. Reviewed your comments and implemented all suggestions
3. Refactored blade code for contact and login forms. Created reusable components
4. Refactored code for validation rules in JS and created a new axios for login validation

_Encountered the following questions:_

As far as I decided to implement client side validation using axios, I found that in case of successful authentication, redirection does not work due to `prevent default` behavior in my front-end management   
```
$credentials = $request->validated();
if (Auth::attempt($credentials)) {
$request->session()->regenerate();
  return redirect('/')->with('status', __('auth.login.success'));
}
```
I came up with the solution to make a manual redirection in my axios. Is it a good practice? I watched my react project and eventually in React I created a personolised hook to redirect a user. Here is my code from the resume project
```
.then((response) => {
   if (response.status === 200) {
   setSuccessCookie();
    window.location.href = "/";
    deleteInputValues(inputs);
    }
})
```
After that redirection started to work but I lost my functionality of a flashed session message. It appears when I log out but not when I log in. I wanted to manually create a cookie with using js and then a middleware to intercept this cookie and make this flashed message work. I didn't manage to work this functionality. I would like to know whether it is a good way to implement this feature. To discuss! 

**Status for 03/07/2024**

1. Find a way around with session flash data. Found `flash` method in the docs https://laravel.com/docs/11.x/session#flash-data that keeps in the session a message for the next http request. Replaced in the back the `redirect()` with `response()->json()` so when in the front I make `window.location.href = "/";` I still have my flashed message in the session. Here is 
in public function store(StoreUserRequest $request). It is funny though that `flash` method is not recognised by VSCode
```
    {
        $credentials = $request->validated();


        if (Auth::attempt($credentials)) {
            $request->session()->regenerate();
            $request->session()->flash('status', __('auth.login.success'));

            return response()->json([], HttpResponse::HTTP_OK);
        }

        return response()->json([
            'status' => false,
            'message' => __('auth.failed'),
        ], HttpResponse::HTTP_BAD_REQUEST);
    }
```
2. Finished installation of Laravel Telescope

### CODE REVIEW

File: app/Http/Controllers/SessionsController.php [hunk 1]
```diff
@@ -12,30 +13,28 @@ public function login()
...
-                'email' => 'Your provided credentials could not be verified'
-            ]);
+ 
+        $credentials = $request->validated();
+        
+        
+        if (Auth::attempt($credentials)) {
+            $request->session()->regenerate();
+ 
+            return redirect('/')->with('status', 'Welcome back');;
```
Comment (stephdotnet) [inline app/Http/Controllers/SessionsController.php:25]:
```suggestion
            return redirect('/')->with('status', __('auth.login.success')');
```

File: app/Http/Controllers/SessionsController.php [hunk 2]
```diff
@@ -12,30 +13,28 @@ public function login()
...
+            'message' => __('auth.failed'),
+        ], HttpResponse::HTTP_BAD_REQUEST);
     }
 
     public function destroy()
     {
         auth()->logout();
 
-        return redirect('/')->with('success', 'Goodbye!');
+        return redirect('/')->with('status', 'Goodbye!');
```
Comment (stephdotnet) [inline app/Http/Controllers/SessionsController.php:38]:
```suggestion
        return redirect('/')->with('status', __('auth.logout.success'));
```

File: app/Http/Middleware/Authenticate.php [hunk 3]
```diff
@@ -12,6 +12,9 @@ class Authenticate extends Middleware
      */
     protected function redirectTo(Request $request): ?string
     {
-        return $request->expectsJson() ? null : route('login');
+        //QUESTION
+        // return $request->expectsJson() ? null : route('login');
+        //Here I prefer to redirect to home if a not authenticated user tries to access the forbidden routes
```
Comment (stephdotnet) [inline app/Http/Middleware/Authenticate.php:17]: Agreed

File: app/Http/Middleware/IsAdmin.php
Comment (stephdotnet) [inline app/Http/Middleware/IsAdmin.php:1]: To remove comme vu ensemble

File: app/Models/Contact.php [hunk 4]
```diff
@@ -18,7 +18,7 @@ public static function boot()
         parent::boot();
 
         static::created(function ($item) {
-            $adminEmail = env('ADMIN_EMAIL');
+            $adminEmail = config('admin.email');
             Mail::to($adminEmail)->send(new ContactMail($item));
```
Comment (stephdotnet) [inline app/Models/Contact.php:22]:
```suggestion
            Mail::to(config('admin.email'))->send(new ContactMail($item));
```

File: app/Providers/AppServiceProvider.php [hunk 5]
```diff
@@ -22,7 +22,7 @@ public function register(): void
     public function boot(): void
     {
         Gate::define('admin', function (User $user) {
-            return $user->username === 'admin';
+            return $user->name === 'admin';
```
Comment (stephdotnet) [inline app/Providers/AppServiceProvider.php:25]: faire avec l'email comme vu ensemble

File: app/Providers/TelescopeServiceProvider.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+use Laravel\Telescope\TelescopeApplicationServiceProvider;
+
+class TelescopeServiceProvider extends TelescopeApplicationServiceProvider
+{
+    /**
+     * Register any application services.
+     */
+    public function register(): void
+    {
+        // Telescope::night();
```
Comment (stephdotnet) [inline app/Providers/TelescopeServiceProvider.php:17]: Soit supprimer, soit décommenter (dark mode de telescope).

Une fois les migrations jouées tu peux accéder à téléscope sur localhost/telescope

File: app/Providers/TelescopeServiceProvider.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace App\Providers;
+
+use Illuminate\Support\Facades\Gate;
+use Laravel\Telescope\IncomingEntry;
+use Laravel\Telescope\Telescope;
+use Laravel\Telescope\TelescopeApplicationServiceProvider;
+
+class TelescopeServiceProvider extends TelescopeApplicationServiceProvider
```
Comment (stephdotnet) [inline app/Providers/TelescopeServiceProvider.php:10]: Tu devrais plutôt suivre l'installation "local only" (pas le meme provider) https://laravel.com/docs/11.x/telescope#local-only-installation

File: database/seeders/DatabaseSeeder.php [hunk 8]
```diff
@@ -12,11 +15,10 @@ class DatabaseSeeder extends Seeder
...
     public function run(): void
     {
-        \App\Models\User::factory()->create([
+        User::factory()->create([
             'name' => 'admin',
-            'username' => 'admin',
-            'email' => env('ADMIN_EMAIL'),
-            'password' => env('ADMIN_PASSWORD'),
+            'email' => config('admin.email'),
+            'password' => Hash::make(config('admin.password')),
```
Comment (stephdotnet) [inline database/seeders/DatabaseSeeder.php:21]: Je suis pas sur que tu doive faire "Hash::make" car tu as déjà cast "password" en "Hashed"

File: resources/views/components/layout.blade.php [hunk 9]
```diff
@@ -15,14 +15,20 @@
...
         </button>
     </div>
 
     {{ $slot }}
+
+    @if (session('status'))
+    <div class="absolute bottom-0 right-0">
+        {{ session('status') }}
+    </div>
+    @endif
```
Comment (stephdotnet) [inline resources/views/components/layout.blade.php:31]: Peut être faire un composant

File: resources/views/sessions/login.blade.php [hunk 10]
```diff
@@ -2,12 +2,37 @@
...
             <h1 class="text-center font-bold text-xl">Log in!</h1>
-            <form method="POST" action="/sessions" class="mt-10">
+            
+            <form method="POST" class="mt-4 relative" action="{{ route('session.store') }}" id="loginForm">
                 @csrf
+                <div class="flex flex-col gap-2">
+                    <div class="sm:col-span-4">
+                        <label for="email"
+                            class="block text-sm font-bold leading-6 text-gray-900">{{ __('form.label.email') }}</label>
+                        {{-- <div data-inputName="email"></div> --}}
```
Comment (stephdotnet) [inline resources/views/sessions/login.blade.php:12]: To remove

File: resources/views/sessions/login.blade.php [hunk 11]
```diff
@@ -2,12 +2,37 @@
...
+                        <div class="mt-2">
+                            <input type="email" name="email" id="email"
+                                class="block w-full rounded border-0 bg-white py-1.5 text-gray-900 ring-gray-300 ring-1 focus:ring-1 focus:ring-inset focus:ring-purple-600 placeholder:text-gray-400 placeholder:text-[14px]"
+                                placeholder="{{ __('form.input.email') }}" value="{{ old('email') }}">
+                        </div>
+                    </div>
+                    <div class="sm:col-span-4">
+                        <label for="password"
+                            class="block text-sm font-bold leading-6 text-gray-900">{{ __('form.label.password') }}</label>
+                        {{-- <div data-inputName="password"></div> --}}
```
Comment (stephdotnet) [inline resources/views/sessions/login.blade.php:22]: To remove

File: resources/views/sessions/login.blade.php [hunk 12]
```diff
@@ -2,12 +2,37 @@
...
+                        <div class="mt-2">
+                            <input type="password" name="password" id="password"
+                                class="block w-full rounded border-0 bg-white py-1.5 text-gray-900 ring-1 ring-gray-300 focus:ring-1 focus:ring-inset focus:ring-purple-600 placeholder:text-gray-400 placeholder:text-[14px]"
+                                placeholder="{{ __('form.input.password') }}">
```
Comment (stephdotnet) [inline resources/views/sessions/login.blade.php:26]: Si c'est le meme style que tu formulaire de contact, tu peux peut-etre faire un composant blade pour les input du formulaire :)

File: routes/web.php [hunk 13]
```diff
@@ -33,12 +35,25 @@
...
-Route::post('admin/photos', [AdminController::class, 'store'])->middleware('can:admin');
-Route::delete('admin/photos/{photo}', [AdminController::class, 'destroy'])->middleware('can:admin');
+Route::post('/contact-us', [ContactController::class, 'store'])->name('contact.us.store');
+Route::get('/login', [SessionsController::class, 'login'])->name('login');
+Route::post('/sessions', [SessionsController::class, 'store'])->name('session.store');;
+
+//QUESTION
+//Is it better my approach with personolised middlewear IsAdmin
+// Or I use Gate approach from AppServiceProvider with can directive? 
+// Route::get('admin', [AdminController::class, 'index'])->middleware('can:admin')->name('admin');
```
Comment (stephdotnet) [inline routes/web.php:45]: Vu ensemble pour faire plutôt `Route::middleware('can:admin')` sur le groupe

**Review by stephdotnet — COMMENTED** [review]
Bravo 🙏👏

For next PR:

- [ ] Install laravel pint
- [ ] Use laravel telecope
- [ ] Setup auto deployment on server

File: app/Http/Middleware/Authenticate.php (hunk 3 above)
Comment (purplena) [reply app/Http/Middleware/Authenticate.php:17]: deleted unnecessary comments

File: app/Http/Middleware/IsAdmin.php
Comment (purplena) [reply app/Http/Middleware/IsAdmin.php:1]: Updated a gate with admin's email
```
Gate::define('admin', function (User $user) {
            return $user->email === config('admin.email');
        });
```

File: app/Providers/TelescopeServiceProvider.php (hunk 6 above)
Comment (purplena) [reply app/Providers/TelescopeServiceProvider.php:17]: uncommented! c'est génial

File: app/Providers/TelescopeServiceProvider.php (hunk 7 above)
Comment (purplena) [reply app/Providers/TelescopeServiceProvider.php:10]: done! now Laravel Telescope is in `require-dev`

File: database/seeders/DatabaseSeeder.php (hunk 8 above)
Comment (purplena) [reply database/seeders/DatabaseSeeder.php:21]: I do not know how you see this kind of things but yes, I removed `Hash `Facade and it worked

File: resources/views/components/layout.blade.php (hunk 9 above)
Comment (purplena) [reply resources/views/components/layout.blade.php:31]: done with added animation and js to hide a flash-message

File: app/Providers/TelescopeServiceProvider.php (hunk 7 above)
Comment (stephdotnet) [reply app/Providers/TelescopeServiceProvider.php:10]: il faut faire tout le chapitre (changer le service provider et rajouter le "dont discover"

File: routes/web.php [hunk 14]
```diff
@@ -32,13 +32,25 @@
...
+
+Route::group(['middleware' => 'auth'], function () {
+    Route::middleware('can:admin')->group(function () {
+        Route::get('admin', [AdminController::class, 'index'])->name('admin');
+        Route::get('photos', [AdminController::class, 'photos'])->name('photos');
+        Route::post('logout', [SessionsController::class, 'destroy'])->withoutMiddleware(['auth']);
+        Route::get('admin/photos/upload', [AdminController::class, 'create'])->middleware('can:admin');
+        Route::post('admin/photos', [AdminController::class, 'store'])->middleware('can:admin');
+        Route::delete('admin/photos/{photo}', [AdminController::class, 'destroy'])->middleware('can:admin');
```
Comment (stephdotnet) [inline routes/web.php:54]: Le groupe est déjà protégé par le middleware `can:admin` donc ce n'est pas nécessaire de l'avoir sur chaque route
```suggestion
        Route::get('admin/photos/upload', [AdminController::class, 'create']);
        Route::post('admin/photos', [AdminController::class, 'store']);
        Route::delete('admin/photos/{photo}', [AdminController::class, 'destroy']);
```

File: routes/web.php [hunk 15]
```diff
@@ -32,13 +32,25 @@
...
+
+Route::post('/contact-us', [ContactController::class, 'store'])->name('contact.us.store');
+Route::get('/login', [SessionsController::class, 'login'])->name('login');
+Route::post('/sessions', [SessionsController::class, 'store'])->name('session.store');
+
+
+//QUESTION
+//Is it better my approach with personolised middlewear IsAdmin
+// Or I use Gate approach from AppServiceProvider with can directive?
+// Route::get('admin', [AdminController::class, 'index'])->middleware('can:admin')->name('admin');
```
Comment (stephdotnet) [inline routes/web.php:45]: to remove

File: app/Http/Middleware/CheckStatusCookie.php [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        if ($request->hasCookie('status')) {
+            $status = $request->cookie('status');
+            if ($status === 'success') {
+                Session::put('status', __('auth.login.success'));
+            }
+        }
+
+        return $next($request);
+    }
```
Comment (stephdotnet) [inline app/Http/Middleware/CheckStatusCookie.php:27]: Je pense que tu as trop complexifié la chose. Tu n'as pas besoin de ce middleware.

Dans ton `app/Http/Controllers/SessionsController.php` tu fais déjà un redirect "with" (donc tu ajoute dans la session "status" ton message)
image

Donc tu n'a pas besoin de le rajouter ici

image

File: resources/js/loginFormValidation.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+                console.error("Error response is missing:", error);
+            }
+        });
+});
+
+function setSuccessCookie() {
+    const fiveSecondsFromNow = new Date();
+    fiveSecondsFromNow.setSeconds(fiveSecondsFromNow.getSeconds() + 35);
+    document.cookie = `status=success; expires=${fiveSecondsFromNow.toUTCString()}; path=/`;
+}
```
Comment (stephdotnet) [inline resources/js/loginFormValidation.js:44]: to remove (le back s'occupe de mettre en session le message de status)

File: resources/js/loginFormValidation.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+    deleteErrorMessages(document.querySelectorAll("[data-inputName]"));
+    deleteRedErrorRing(inputs);
+
+    axios
+        .post("/sessions", new FormData(this))
+        .then((response) => {
+            if (response.status === 200) {
+                setSuccessCookie();
+                window.location.href = "/";
+                deleteInputValues(inputs);
```
Comment (stephdotnet) [inline resources/js/loginFormValidation.js:25]:
```suggestion
                window.location.href = "/";
                deleteInputValues(inputs);
```

File: resources/views/components/sections/components/flashed-session.blade.php [hunk 19]
```diff
@@ -0,0 +1,6 @@
+{{-- @if (session('status')) --}}
```
Comment (stephdotnet) [inline resources/views/components/sections/components/flashed-session.blade.php:1]: to remove

**Review by stephdotnet — APPROVED** [review]
good, mais il faut qu'on parle de la partie "message" de confirmation

File: app/Http/Middleware/CheckStatusCookie.php (hunk 16 above)
Comment (purplena) [reply app/Http/Middleware/CheckStatusCookie.php:27]: problem resolved

**Pull request merged**